### ISSUE TITLE
Non-blocking rwops

### ISSUE THREAD

**Issue by icculus** [libsdl-org/SDL#6820]
While I'm getting medieval on the RWops API, I figured I should open this up for discussion:

The changes I made yesterday change the return value of RWread to something like this:

- a positive number: number of bytes read
- zero: eof
- negative 1: error (check SDL_GetError for human-readable details)
- **negative 2**: this is a non-blocking RWops and we would have to wait, so try again later.

(RWwrite is now positive number of bytes written, or -1 on error when no bytes were written. If you wrote less than expected, it's either non-blocking or the next call is likely to error out.)

Both of these functions were not renamed, but now take one less argument, so existing apps will have to examine these calls.

This is all a little unorthodox, but my thinking is that a) things that don't handle non-blocking reads will see a negative value and treat it as an error and b) POSIX gives you an actual error (-1) and wants you to check errno for EWOULDBLOCK, but checking SDL_GetError for this is messier for several reasons.

There are currently no non-blocking RWops offered by SDL, and I will not add a way to _toggle_ non-blocking state. I think eventually it would be new functions (SDL_RWFromFileNonblocking or whatever) to create a non-blocking RWops. The goal is to make this flexible enough that apps and libraries implement their own RWops that hook up to raw sockets, HTTP transfers, etc, though.

Opinions and arguments welcome! This can still be changed.

**Comment by slouken** [comment]
I like it!

While we're at it, does it make sense to have a callback based I/O completion port style API?

**Comment by icculus** [comment]
Maybe a callback is provided when creating a non-blocking RWops and it fires in addition to the usual non-blocking semantics? So you can optionally use it to notify your app so you don't have to poll, but it isn't a formal part of the RWops struct...that one internal implementation deals with it and not everyone that implements a RWops...

**Comment by slouken** [comment]
Callback based I/O is often used for asynchronous asset loading from disk, not just for non-blocking network I/O. It's also the most efficient way to handle network I/O on Linux.

**Comment by 1bsyl** [comment]
probably not better but other ideas:
- could be also an event based, async API would push back some SDL Event ?
- could also be done natively using some Rust layer  ? 
- provide some kind of template, so that any API can be turned into an async command

**Comment by nfries88** [comment]
Hi, I'm the one who originally proposed an async RWops API way back when SDL2 was new

non-blocking file I/O doesn't really work when the file is on a local filesystem. Those files are always "ready to read" and "ready to write" as far as the OS is concerned. This is true whether you're using O_NONBLOCK, poll, select, epoll, kqueue, etc. Those were designed for "files" with unreliable latencies in mind - pipes, sockets, character terminals, etc - not the local filesystem.

This is mostly mostly only really important when doing non-sequential reads, because the OS already buffers writes for you and readahead buffering is **generally** good enough to keep the data flowing for sequential reads. When the data isn't buffered already, now you're stuck waiting on the disk, and that can be a problem.

So synchronous non-blocking I/O like on a socket or a pipe just doesn't work. The only solution is async I/O. You probably want a random-access interface, not a streaming interface like the current RWops. You may even want a vectored interface (like libuv's uv_fs_read and uv_fs_write) but not sure how well used that particularly functionality would be.

I actually worked on a small async disk file I/O interface for SDL programs a couple years ago, but I kinda lost interest pretty quick. Source code for the marginally broader library is here if you want to give it a look: https://sourceforge.net/p/libs4sdl2/code/ci/default/tree/SDL_EXT/

Callbacks or pushing an event into the queue both work fine, but for C++ devs callbacks will be nicer (coroutines and lamdas). I definitely wouldn't do futures.

**Comment by slouken** [comment]
Games are always able to create work threads and do loading I/O on those threads. What are the use cases where this provides meaningful benefit for the application?

**Comment by icculus** [comment]
The thinking is "real" async, at the kernel layer, can be implemented, but things that can't implement it have a functionally equivalent fallback.

But maybe the real async interfaces aren't worth more than just spinning a thread anyhow...?

**Comment by slouken** [comment]
The only case I know where async I/O is critical is high performance networking. In that case you'd want to use native interfaces directly.

Maybe I'm missing something?

**Comment by icculus** [comment]
I think this is less performance and more "I want data from the disk when it's available, and I don't want to block _at all_ while the data is coming off the disk/competing with other things that want the disk, etc".

Two things I can think of immediately is a loading screen that keeps animating at 60 fps, because it's able to deal with chunks of data for the first time when they are in RAM without having to wait for a read to finish, because they were notified when it was ready, and open world games that are sort of continuously streaming data from the disk as you drive around the world.

This is a different thing than non-blocking i/o, since those facilities (select(), poll(), etc) always treat the local disk as "ready," even if the disk needs to spin up from low-power mode. Win32 calls this thing "overlapped i/o".

(But again, maybe this isn't worth it. I'm just proposing a solution to an open bug, but closing it as WONTFIX is a solution too.  :)  )

**Comment by slouken** [comment]
Yeah, for that use case this seems like a good addition to the API.

**Comment by nfries88** [comment]
> But maybe the real async interfaces aren't worth more than just spinning a thread anyhow...?

On Windows overlapped I/O is tightly integrated with the filesystem cache, so there is a pretty significant advantage in terms of both system resource use and general overhead. FreeBSD also stands out with direct kernel support for the posix AIO interface, plus an extension to get notification via kqueue instead of signals.

Linux has had a wonky kernel AIO interface *(not used by the posix AIO implementation, which is an incredibly naiive threadpool in userspace)* that has been favored by databases for a long time but probably wouldn't be advantageous to most software using SDL, and if the security vulnerabilities ever get resolved clever use of io_uring features provide notable performance advantages for modest reads over many files but it would probably be challenging to encapsulate that use pattern in a "simple" general purpose interface.

emscripten also stands out because it has asynchronous filesystem APIs (even though they're ugly as sin) and because it is probably the biggest (or only?) target for SDL where threads may not be an option.

**Comment by icculus** [comment]
Okay, I've redesigned this interface three times in three days, and I still hate where it's at.

It's become clear to me that this shouldn't be added to a generic SDL_IOStream; my plan was pretty naive. It should probably be a totally separate API if we do it at all, but I haven't found a way to make something that's portable, flexible, useful, and pleasant to use, at least so far.

But if we're not putting it into SDL_IOStream, we can bump this out of the 3.0 ABI milestone, and maybe revisit it later.

For now, I'm moving to 3.2.0, but honestly I might bump it to 3.x. Or give up on it.

**Comment by nfries88** [comment]
I would probably make it part of the Storage API tbh, async I/O from cloud and optical disk storage would probably be nice to have too.

async cancellability is inherently unreliable, adds a little overhead, and for local filesystem I/O is probably completely pointless unless the game is loading massive amounts of data from more files than the system can handle at once or the system actually supports I/O cancellation after initiation (posix AIO supports cancellation in its interface, but in practice I don't think any systems will stop in the middle of I/O). 

I know it's something async programmers coming from JS or another highly asynchronous language would expect, but it's probably not worth the effort of supporting.

If dealing with callbacks is part of the headache, putting a completion event in the queue with some opaque user-provided token (as void*) is a good enough approach.

**Comment by icculus** [comment]
Let's track this over in #1374.
